File: nim/src/help_tree.nim
import std/[json, os, strutils, terminal]

type
  TextEmphasis* = enum Normal, Bold, Italic, BoldItalic

  TextTokenTheme* = object
    emphasis*: TextEmphasis
    color_hex*: string

  HelpTreeTheme* = object
    command*: TextTokenTheme
    options*: TextTokenTheme
    description*: TextTokenTheme

  HelpTreeOutputFormat* = enum Text, Json
  HelpTreeStyle* = enum Plain, Rich
  HelpTreeColor* = enum Auto, Always, Never

  HelpTreeOpts* = object
    depth_limit*: int
    ignore*: seq[string]
    tree_all*: bool
    output*: HelpTreeOutputFormat
    style*: HelpTreeStyle
    color*: HelpTreeColor
    theme*: HelpTreeTheme

  HelpTreeInvocation* = object
    opts*: HelpTreeOpts
    path*: seq[string]

  HelpTreeConfigFile* = object
    theme*: HelpTreeTheme

proc defaultTheme*(): HelpTreeTheme =
  result.command = TextTokenTheme(emphasis: Bold, color_hex: "#7ee7e6")
  result.options = TextTokenTheme(emphasis: Normal)
  result.description = TextTokenTheme(emphasis: Italic, color_hex: "#90a2af")

proc defaultOpts*(): HelpTreeOpts =
  result.output = Text
  result.style = Rich
  result.color = Auto
  result.theme = defaultTheme()

proc shouldUseColor*(opts: HelpTreeOpts): bool =
  case opts.color
  of Always: true
  of Never: false
  of Auto: stdout.isatty

proc parseHexRGB*(hex: string): tuple[r, g, b: int] =
  let h = hex.strip(chars = {'#'})
  if h.len == 6:
    result.r = parseHexInt(h[0..1])
    result.g = parseHexInt(h[2..3])
    result.b = parseHexInt(h[4..5])

proc styleText*(text: string, token: TextTokenTheme, opts: HelpTreeOpts): string =
  if opts.style == Plain or (token.emphasis == Normal and token.color_hex.len == 0):
    return text
  var codes: seq[string]
  case token.emphasis
  of Bold: codes.add("1")
  of Italic: codes.add("3")
  of BoldItalic: codes.add("1"); codes.add("3")
  of Normal: discard
  if shouldUseColor(opts) and token.color_hex.len > 0:
    let rgb = parseHexRGB(token.color_hex)
    codes.add("38;2;" & $rgb.r & ";" & $rgb.g & ";" & $rgb.b)
  if codes.len == 0: return text
  result = "\x1b[" & codes.join(";") & "m" & text & "\x1b[0m"

proc loadConfig*(path: string): HelpTreeConfigFile =
  let data = readFile(path)
  let j = parseJson(data)
  if j.hasKey("theme"):
    let t = j["theme"]
    proc parseToken(node: JsonNode): TextTokenTheme =
      if node.hasKey("emphasis"):
        result.emphasis = parseEnum[TextEmphasis](node["emphasis"].getStr)
      if node.hasKey("color_hex"):
        result.color_hex = node["color_hex"].getStr
    result.theme.command = parseToken(t["command"])
    result.theme.options = parseToken(t["options"])
    result.theme.description = parseToken(t["description"])

proc applyConfig*(opts: var HelpTreeOpts, config: HelpTreeConfigFile) =
  opts.theme = config.theme

proc parseHelpTreeInvocation*(argv: seq[string]): HelpTreeInvocation =
  var helpTree = false
  var depthLimit = -1
  var ignore: seq[string]
  var treeAll = false
  var output = Text
  var style = Rich
  var color = Auto
  var path: seq[string]

  var i = 0
  while i < argv.len:
    let arg = argv[i]
    case arg
    of "--help-tree": helpTree = true
    of "--tree-depth", "-L":
      i += 1
      depthLimit = parseInt(argv[i])
    of "--tree-ignore", "-I":
      i += 1
      ignore.add(argv[i])
    of "--tree-all", "-a": treeAll = true
    of "--tree-output":
      i += 1
      output = parseEnum[HelpTreeOutputFormat](argv[i].capitalizeAscii)
    of "--tree-style":
      i += 1
      style = parseEnum[HelpTreeStyle](argv[i].capitalizeAscii)
    of "--tree-color":
      i += 1
      color = parseEnum[HelpTreeColor](argv[i].capitalizeAscii)
    else:
      if not arg.startsWith("-"):
        path.add(arg)
    i += 1

  if not helpTree:
    return HelpTreeInvocation(opts: defaultOpts(), path: @[])

  var opts = defaultOpts()
  if depthLimit >= 0: opts.depth_limit = depthLimit
  opts.ignore = ignore
  opts.tree_all = treeAll
  opts.output = output
  opts.style = style
  opts.color = color
  return HelpTreeInvocation(opts: opts, path: path)

proc runForParser*(parser: auto, opts: HelpTreeOpts, requestedPath: seq[string] = @[]) =
  # Simplified for cligen; actual implementation would introspect cligen AST
  echo styleText("myapp", opts.theme.command, opts)
  echo ""
  echo "Use `myapp <COMMAND> --help` for full details on arguments and flags."

File: zig/src/help_tree.zig
const std = @import("std");

pub const TextEmphasis = enum { normal, bold, italic, bold_italic };

pub const TextTokenTheme = struct {
    emphasis: TextEmphasis = .normal,
    color_hex: ?[]const u8 = null,
};

pub const HelpTreeTheme = struct {
    command: TextTokenTheme = .{ .emphasis = .bold, .color_hex = "#7ee7e6" },
    options: TextTokenTheme = .{},
    description: TextTokenTheme = .{ .emphasis = .italic, .color_hex = "#90a2af" },
};

pub const HelpTreeOutputFormat = enum { text, json };
pub const HelpTreeStyle = enum { plain, rich };
pub const HelpTreeColor = enum { auto, always, never };

pub const HelpTreeOpts = struct {
    depth_limit: ?usize = null,
    ignore: []const []const u8 = &.{},
    tree_all: bool = false,
    output: HelpTreeOutputFormat = .text,
    style: HelpTreeStyle = .rich,
    color: HelpTreeColor = .auto,
    theme: HelpTreeTheme = .{},
};

pub const HelpTreeInvocation = struct {
    opts: HelpTreeOpts,
    path: []const []const u8,
};

fn shouldUseColor(opts: HelpTreeOpts) bool {
    return switch (opts.color) {
        .always => true,
        .never => false,
        .auto => std.io.getStdOut().isTty(),
    };
}

fn parseHexRgb(hex: []const u8) ?struct { r: u8, g: u8, b: u8 } {
    const h = std.mem.trimLeft(u8, hex, "#");
    if (h.len != 6) return null;
    const r = std.fmt.parseInt(u8, h[0..2], 16) catch return null;
    const g = std.fmt.parseInt(u8, h[2..4], 16) catch return null;
    const b = std.fmt.parseInt(u8, h[4..6], 16) catch return null;
    return .{ .r = r, .g = g, .b = b };
}

pub fn styleText(allocator: std.mem.Allocator, text: []const u8, token: TextTokenTheme, opts: HelpTreeOpts) ![]const u8 {
    if (opts.style == .plain or (token.emphasis == .normal and token.color_hex == null))
        return allocator.dupe(u8, text);

    var codes: [4]u8 = undefined;
    var code_count: usize = 0;
    switch (token.emphasis) {
        .bold => {
            codes[code_count] = 1;
            code_count += 1;
        },
        .italic => {
            codes[code_count] = 3;
            code_count += 1;
        },
        .bold_italic => {
            codes[code_count] = 1;
            code_count += 1;
            codes[code_count] = 3;
            code_count += 1;
        },
        .normal => {},
    }

    if (shouldUseColor(opts)) {
        if (token.color_hex) |hex| {
            if (parseHexRgb(hex)) |rgb| {
                var buf: [32]u8 = undefined;
                const s = try std.fmt.bufPrint(&buf, "38;2;{d};{d};{d}", .{ rgb.r, rgb.g, rgb.b });
                _ = s;
            }
        }
    }

    if (code_count == 0) return allocator.dupe(u8, text);

    var result = std.ArrayList(u8).init(allocator);
    defer result.deinit();
    try result.appendSlice("\x1b[");
    for (codes[0..code_count], 0..) |code, i| {
        if (i > 0) try result.append(';');
        try result.writer().print("{d}", .{code});
    }
    try result.append('m');
    try result.appendSlice(text);
    try result.appendSlice("\x1b[0m");
    return result.toOwnedSlice();
}

pub fn hasHelpTree(argv: []const []const u8) bool {
    for (argv) |arg| {
        if (std.mem.eql(u8, arg, "--help-tree")) return true;
    }
    return false;
}

pub fn parseInvocation(allocator: std.mem.Allocator, argv: []const []const u8) !?HelpTreeInvocation {
    var help_tree = false;
    var depth_limit: ?usize = null;
    var ignore = std.ArrayList([]const u8).init(allocator);
    defer ignore.deinit();
    var tree_all = false;
    var output: HelpTreeOutputFormat = .text;
    var style: HelpTreeStyle = .rich;
    var color: HelpTreeColor = .auto;
    var path = std.ArrayList([]const u8).init(allocator);
    defer path.deinit();

    var i: usize = 0;
    while (i < argv.len) : (i += 1) {
        const arg = argv[i];
        if (std.mem.eql(u8, arg, "--help-tree")) {
            help_tree = true;
        } else if (std.mem.eql(u8, arg, "--tree-depth") or std.mem.eql(u8, arg, "-L")) {
            i += 1;
            if (i >= argv.len) return error.MissingValue;
            depth_limit = try std.fmt.parseInt(usize, argv[i], 10);
        } else if (std.mem.eql(u8, arg, "--tree-ignore") or std.mem.eql(u8, arg, "-I")) {
            i += 1;
            if (i >= argv.len) return error.MissingValue;
            try ignore.append(argv[i]);
        } else if (std.mem.eql(u8, arg, "--tree-all") or std.mem.eql(u8, arg, "-a")) {
            tree_all = true;
        } else if (std.mem.eql(u8, arg, "--tree-output")) {
            i += 1;
            if (i >= argv.len) return error.MissingValue;
            output = std.meta.stringToEnum(HelpTreeOutputFormat, argv[i]) orelse return error.InvalidValue;
        } else if (std.mem.eql(u8, arg, "--tree-style")) {
            i += 1;
            if (i >= argv.len) return error.MissingValue;
            style = std.meta.stringToEnum(HelpTreeStyle, argv[i]) orelse return error.InvalidValue;
        } else if (std.mem.eql(u8, arg, "--tree-color")) {
            i += 1;
            if (i >= argv.len) return error.MissingValue;
            color = std.meta.stringToEnum(HelpTreeColor, argv[i]) orelse return error.InvalidValue;
        } else if (!std.mem.startsWith(u8, arg, "-")) {
            try path.append(arg);
        }
    }

    if (!help_tree) return null;

    return HelpTreeInvocation{
        .opts = .{
            .depth_limit = depth_limit,
            .ignore = try ignore.toOwnedSlice(),
            .tree_all = tree_all,
            .output = output,
            .style = style,
            .color = color,
        },
        .path = try path.toOwnedSlice(),
    };
}
